sklearn/preprocessing/tests/test_data.py

```diff
@@ -851,6 +853,291 @@ def test_robust_scaler_iris_quantiles():
+
+    # consider the case where sparse entries are missing values and user-given
+    # zeros are to be considered
+    X_data = np.array([0, 0, 1, 0, 2, 2, 1, 0, 1, 2, 0])
+    X_col = np.array([0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1])
+    X_row = np.array([0, 4, 0, 1, 2, 3, 4, 5, 6, 7, 8])
+    X_sparse = sparse.csc_matrix((X_data, (X_row, X_col)))
+    X_trans = transformer.fit_transform(X_sparse)
+    X_gt = np.array([[0., 0.5],
+                     [0., 0.],
+                     [0., 1.],
+                     [0., 1.],
+                     [0., 0.5],
+                     [0., 0.],
+                     [0., 0.5],
+                     [0., 1.],
+                     [0., 0.]])
+    assert_almost_equal(X_gt, X_trans.A)
+
+    transformer = QuantileTransformer(ignore_implicit_zeros=True,
+                                      n_quantiles=5)
+    X_data = np.array([-1, -1, 1, 0, 0, 0, 1, -1, 1])
+    X_col = np.array([0, 0, 1, 1, 1, 1, 1, 1, 1])
+    X_row = np.array([0, 4, 0, 1, 2, 3, 4, 5, 6])
+    X_sparse = sparse.csc_matrix((X_data, (X_row, X_col)))
+    X_trans = transformer.fit_transform(X_sparse)
+    X_gt = np.array([[0, 1],
+                     [0, 0.5],
+                     [0, 0.5],
+                     [0, 0.5],
+                     [0, 1],
+                     [0, 0],
+                     [0, 1]])
+    assert_almost_equal(X_gt, X_trans.A)
+    assert_almost_equal(X_sparse.A, transformer.inverse_transform(X_trans).A)
+
+
+def test_quantile_transform_dense_toy():
+    X = np.array([[0, 25, 50, 75, 100],
+                  [2, 4, 6, 8, 10],
+                  [2.6, 4.1, 2.3, 9.5, 0.1]]).T
+
+    transformer = QuantileTransformer(n_quantiles=5)
+    transformer.fit(X)
+
+    X_trans = transformer.fit_transform(X)
+    X_gt = np.tile(np.linspace(0, 1, num=5), (3, 1)).T
+    assert_almost_equal(np.sort(X_trans, axis=0), X_gt)
+
+    X_test = np.array([
+        [-1,  1,  0],
+        [101, 11, 10],
+    ])
+    expected = np.array([
+        [0, 0, 0],
+        [1, 1, 1],
+    ])
+    assert_array_almost_equal(transformer.transform(X_test), expected)
+
+    X_trans_inv = transformer.inverse_transform(X_trans)
+    assert_array_almost_equal(X, X_trans_inv)
+
+
+def test_quantile_transform_subsampling():
+    # dense support
+    N = 1000000
+    X = np.sort(np.random.sample((N, 1)), axis=0)
+    ROUND = 5
+    inf_norm_arr = []
+    for random_state in range(ROUND):
+        transformer_subsample = QuantileTransformer(random_state=random_state,
+                                                    n_quantiles=N,
+                                                    subsample=N // 10)
+        transformer_subsample.fit(X)
+        inf_norm = np.max(np.abs(np.linspace(0, 1, N) -
+                                 np.ravel(transformer_subsample.quantiles_)))
+        assert_true(inf_norm < 1e-2)
+        inf_norm_arr.append(inf_norm)
+    assert_equal(len(np.unique(inf_norm_arr)), len(inf_norm_arr))
```

Comment:
Add a comment such that: "each random subsampling yield a unique approximation to the expected linspace CDF".